automata/fa/fa.py

```diff
@@ -1,19 +1,233 @@
+            - path (str or os.PathLike, optional): Path to output file. If
+              None, the output will not be saved.
+            - horizontal (bool, optional): Direction of node layout. Defaults
+              to True.
+            - reverse_orientation (bool, optional): Reverse direction of node
+              layout. Defaults to False.
+            - fig_size (tuple, optional): Figure size. Defaults to None.
+            - font_size (float, optional): Font size. Defaults to 14.0.
+            - arrow_size (float, optional): Arrow head size. Defaults to 0.85.
+            - state_separation (float, optional): Node distance. Defaults to 0.5.
+        Returns:
+            AGraph corresponding to the given automaton.
+        """
+
+        if not _visual_imports:
+            raise ImportError(
+                "Missing visualization packages; "
+                "please install coloraide and pygraphviz."
+            )
+
+        # Defining the graph.
+        graph = pgv.AGraph(strict=False, directed=True)
+
+        if fig_size is not None:
+            graph.graph_attr.update(size=", ".join(map(str, fig_size)))
+
+        graph.graph_attr.update(ranksep=str(state_separation))
+        font_size_str = str(font_size)
+        arrow_size_str = str(arrow_size)
+
+        if horizontal:
+            rankdir = "RL" if reverse_orientation else "LR"
+        else:
+            rankdir = "BT" if reverse_orientation else "TB"
+
+        graph.graph_attr.update(rankdir=rankdir)
+
+        # we use a random uuid to make sure that the null node has a
+        # unique id to avoid colliding with other states.
+        # To be able to set the random seed, took code from:
+        # https://nathanielknight.ca/articles/consistent_random_uuids_in_python.html
+        null_node = str(
+            uuid.UUID(bytes=bytes(random.getrandbits(8) for _ in range(16)), version=4)
+        )
+        graph.add_node(
+            null_node,
+            label="",
+            tooltip=".",
+            shape="point",
+            fontsize=font_size_str,
+        )
+        initial_node = self.get_state_name(self.initial_state)
+        graph.add_edge(
+            null_node,
+            initial_node,
+            tooltip="->" + initial_node,
+            arrowsize=arrow_size_str,
+        )
+
+        nonfinal_states = map(self.get_state_name, self.states - self.final_states)
+        final_states = map(self.get_state_name, self.final_states)
+        graph.add_nodes_from(nonfinal_states, shape="circle", fontsize=font_size_str)
+        graph.add_nodes_from(final_states, shape="doublecircle", fontsize=font_size_str)
+
+        is_edge_drawn = defaultdict(lambda: False)
+        if input_str is not None:
+            input_path, is_accepted = self._get_input_path(input_str=input_str)
+
+            start_color = coloraide.Color("#ff0")
+            end_color = (
+                coloraide.Color("#0f0") if is_accepted else coloraide.Color("#f00")
```

Comment:
Is there any reason to use coloraide specifically? I think the one feature we use is the interpolation, and colours has this while also having more active support (I think).

Reply:
After taking a look, it seems like colour is pretty light on features and doesn't have type annotations set up. I think sticking with coloraide is actually the way to go.